### ISSUE TITLE
NullPointerException: InvocationTargetException

### ISSUE THREAD

**Issue by chunwei** [vsch/idea-multimarkdown#20]
###WebStorm 10.0.4 ,  windows 10  x64,  JDK 1.8.0_60
##Detail:
InvocationTargetException
java.lang.NullPointerException
	at com.sun.glass.ui.Application.run(Application.java:146)
	at com.sun.javafx.tk.quantum.QuantumToolkit.startup(QuantumToolkit.java:263)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:211)
	at javafx.embed.swing.JFXPanel.initFx(JFXPanel.java:254)
	at javafx.embed.swing.JFXPanel.<init>(JFXPanel.java:269)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditor$MyJFXPanel.<init>(MultiMarkdownFxPreviewEditor.java:221)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditor$MyJFXPanel.<init>(MultiMarkdownFxPreviewEditor.java:221)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditor.<init>(MultiMarkdownFxPreviewEditor.java:291)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditorProvider.createEditor(MultiMarkdownFxPreviewEditorProvider.java:127)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditorProvider.createEditor(MultiMarkdownFxPreviewEditorProvider.java:105)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl$12.run(FileEditorManagerImpl.java:859)
	at com.intellij.util.ui.UIUtil.invokeAndWaitIfNeeded(UIUtil.java:2369)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl.openFileImpl4(FileEditorManagerImpl.java:842)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl.openFileImpl3(FileEditorManagerImpl.java:776)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl$9.run(FileEditorManagerImpl.java:756)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:124)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:99)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:85)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl.openFileImpl2(FileEditorManagerImpl.java:753)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl.openFileWithProviders(FileEditorManagerImpl.java:694)
	at com.intellij.openapi.fileEditor.ex.FileEditorManagerEx.openFile(FileEditorManagerEx.java:146)
	at com.intellij.codeInsight.navigation.NavigationUtil.a(NavigationUtil.java:191)
	at com.intellij.codeInsight.navigation.NavigationUtil.openFileWithPsiElement(NavigationUtil.java:167)
	at com.intellij.ide.projectView.impl.nodes.AbstractPsiBasedNode.navigate(AbstractPsiBasedNode.java:207)
	at com.intellij.ide.projectView.impl.nodes.AbstractPsiBasedNode.navigate(AbstractPsiBasedNode.java:217)
	at com.intellij.ide.projectView.impl.nodes.PsiFileNode.navigate(PsiFileNode.java:132)
	at com.intellij.util.OpenSourceUtil.navigate(OpenSourceUtil.java:53)
	at com.intellij.util.OpenSourceUtil.openSourcesFrom(OpenSourceUtil.java:31)
	at com.intellij.util.EditSourceOnDoubleClickHandler$TreeMouseListener.processDoubleClick(EditSourceOnDoubleClickHandler.java:130)
	at com.intellij.util.EditSourceOnDoubleClickHandler$TreeMouseListener.onDoubleClick(EditSourceOnDoubleClickHandler.java:122)
	at com.intellij.ui.DoubleClickListener.onClick(DoubleClickListener.java:30)
	at com.intellij.ui.ClickListener$1.mouseReleased(ClickListener.java:73)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:290)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	at java.awt.Component.processMouseEvent(Component.java:6525)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at com.intellij.ui.treeStructure.Tree.processMouseEvent(Tree.java:421)
	at com.intellij.ide.dnd.aware.DnDAwareTree.processMouseEvent(DnDAwareTree.java:53)
	at java.awt.Component.processEvent(Component.java:6290)
	at java.awt.Container.processEvent(Container.java:2234)
	at java.awt.Component.dispatchEventImpl(Component.java:4881)
	at java.awt.Container.dispatchEventImpl(Container.java:2292)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4898)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4533)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4462)
	at java.awt.Container.dispatchEventImpl(Container.java:2278)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:751)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:702)
	at java.awt.EventQueue$3.run(EventQueue.java:696)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:721)
	at com.intellij.ide.IdeEventQueue.e(IdeEventQueue.java:734)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:565)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

**Comment by vsch** [comment]
Thanks for the bug report @chunwei. It seems that the JavaFX initialization is failing. It looks like something internal to JavaFX. I will try to duplicate it in my system. For now you should set the Use Old Preview option in settings so you don't get the error every time. If it fails it falls back on using the old preview.

<img width="545" alt="screen shot 2015-09-08 at 1 46 46 am" src="https://cloud.githubusercontent.com/assets/10299086/9727316/9fdb56bc-55cb-11e5-9676-4e464984805a.png">

I'll work on it and let you know if I can find the cause.

**Comment by chunwei** [comment]
@vsch thanks, "Use Old Preview " works for me.  Hope you can fix it soon.

**Comment by vsch** [comment]
When I try to run it with JDK 1.8.60 I also get an error except with a debugger there is a bit more to go on:

```
Sep 08, 2015 1:56:00 AM com.sun.javafx.css.StyleConverterImpl getInstance
SEVERE: StyleConverterImpl : converter Class is null for : com.sun.javafx.scene.layout.region.CornerRadiiConverter
Sep 08, 2015 1:56:00 AM com.sun.javafx.css.StyleConverterImpl readBinary
SEVERE: could not deserialize com.sun.javafx.scene.layout.region.CornerRadiiConverter
could not deserialize com.sun.javafx.scene.layout.region.CornerRadiiConverter
```

It seems JDK 1.8.60 has a problem with something in the libraries for JavaFX. If you have 1.8.51 and are willing to try running WebStorm with that as the boot JDK.

I remember one of the EAP releases of IntelliJ had 1.8.60 bundled but the next one reverted to 1.8.51 and has stayed at this version since. I'll see if it is a know issue with that JDK release.

**Comment by vsch** [comment]
Just tried beta 1.8.66 and getting the same error. Can you try the previous JDK 1.8.51 and see if you get the same problem?

It took hours to figure out how to make it work on IDEA's and I have no clue to what has changed in update 60 but it is internal to the JDK and looks like it is in the native libraries because I am getting a failed core dump.

**Comment by chunwei** [comment]
I tried install JDK 8u51, and set JAVA_HOME,  WEBIDE_JDK and  WEBIDE_JDK_64 to the 8u51 path.

restart Webstorm, uncheck "Use Old Preview ", open a .md file, still get the same error.

after check "Use Old Preview ", the error no longer appears.

**Comment by chunwei** [comment]
maybe the following static methed return null ?
```classConstructor = MultiMarkdownFxPreviewEditor.getConstructor(Project.class, Document.class, boolean.class);```
line:   MultiMarkdownFxPreviewEditorProvider.java:127

**Comment by vsch** [comment]
It is failing inside JavaFX initialization. I will have to try to figure out if there is something different for 1.8u60 that must be done when using JavaFX in swing.

**Comment by ghostsquad** [comment]
:+1: yep, I'm also receiving this error. I can confirm that checking "Use Old Preview" in Settings > Other Settings > MultiMarkdown, makes this error go away.

**Comment by jdelic** [comment]
I'm seeing the same issue with a slightly different stack trace when I'm editing MarkDown files in my project using IntelliJ IDEA 14 running on Windows 10.

```
InvocationTargetException
java.lang.NullPointerException
	at com.sun.glass.ui.Application.run(Application.java:146)
	at com.sun.javafx.tk.quantum.QuantumToolkit.startup(QuantumToolkit.java:263)
	at com.sun.javafx.application.PlatformImpl.startup(PlatformImpl.java:211)
	at javafx.embed.swing.JFXPanel.initFx(JFXPanel.java:254)
	at javafx.embed.swing.JFXPanel.<init>(JFXPanel.java:269)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditor$MyJFXPanel.<init>(MultiMarkdownFxPreviewEditor.java:221)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditor$MyJFXPanel.<init>(MultiMarkdownFxPreviewEditor.java:221)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditor.<init>(MultiMarkdownFxPreviewEditor.java:291)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditorProvider.createEditor(MultiMarkdownFxPreviewEditorProvider.java:127)
	at com.vladsch.idea.multimarkdown.editor.MultiMarkdownFxPreviewEditorProvider.createEditor(MultiMarkdownFxPreviewEditorProvider.java:105)
	at com.intellij.openapi.fileEditor.impl.FileEditorManagerImpl$12.run(FileEditorManagerImpl.java:859)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:749)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:702)
	at java.awt.EventQueue$3.run(EventQueue.java:696)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:719)
	at com.intellij.ide.IdeEventQueue.e(IdeEventQueue.java:734)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:569)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at com.intellij.ide.IdeEventQueue.pumpEventsForHierarchy(IdeEventQueue.java:818)
	at com.intellij.openapi.progress.util.ProgressWindow.startBlocking(ProgressWindow.java:197)
	at com.intellij.openapi.application.impl.ApplicationImpl.runProcessWithProgressSynchronously(ApplicationImpl.java:633)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressSynchronously(CoreProgressManager.java:363)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.runProcessWithProgressSynchronously(ProgressManagerImpl.java:76)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressSynchronously(CoreProgressManager.java:240)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressSynchronously(CoreProgressManager.java:193)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.openProject(ProjectManagerImpl.java:475)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.loadAndOpenProject(ProjectManagerImpl.java:562)
	at com.intellij.ide.impl.ProjectUtil.openProject(ProjectUtil.java:182)
	at com.intellij.ide.RecentProjectsManagerImpl.doOpenProject(RecentProjectsManagerImpl.java:46)
	at com.intellij.ide.ReopenProjectAction.actionPerformed(ReopenProjectAction.java:61)
	at com.intellij.openapi.wm.impl.welcomeScreen.RecentProjectPanel$2.onClick(RecentProjectPanel.java:120)
	at com.intellij.ui.ClickListener$1.mouseReleased(ClickListener.java:73)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:290)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	at java.awt.Component.processMouseEvent(Component.java:6525)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6290)
	at java.awt.Container.processEvent(Container.java:2234)
	at java.awt.Component.dispatchEventImpl(Component.java:4881)
	at java.awt.Container.dispatchEventImpl(Container.java:2292)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4898)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4533)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4462)
	at java.awt.Container.dispatchEventImpl(Container.java:2278)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:751)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:702)
	at java.awt.EventQueue$3.run(EventQueue.java:696)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:721)
	at com.intellij.ide.IdeEventQueue.e(IdeEventQueue.java:734)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:565)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

I first stumbled over [this discussion on JetBrains DevNet](https://devnet.jetbrains.com/message/5557600?tstart=0) while googling the issue and it seems that even though I have an Oracle Java SDK installed (build 1.8.0_51-b16) , IntelliJ IDEA uses the JRE included with its WIndows installer (build 1.8.0_40-b26), so that might have something to do with me seeing the above exception?

Let me know if I can help or test anything.

**Comment by vsch** [comment]
@jdelic, yes your issue is the same. It is the initialization of JavaFX that is failing. However, you are the first to shed some light on it. If the IDEA runs on the embedded JDK then it is expected. I had that problem with PhpStorm 9.5 EAP. I would change its JDK through its change boot JDK GUI and after 3 hours noticed that it did not change the JDK it used. I figured out that to change the boot jdk for it you need to add a `phpstorm.jdk` file (For IDEA the file name is `idea.jdk`) with the path to the alternate jdk, for example on the Mac it is something like:
 
    /Library/Java/JavaVirtualMachines/jdk1.8.0_51.jdk

to the `/Library/Preferences/WebIde95` directory. Restart your PhpStorm. 

But I don't know if it will work for earlier releases, for these you can refer to the instructions [Selecting the JDK version the IDE will run under](https://intellij-support.jetbrains.com/hc/en-us/articles/206827547-Selecting-the-JDK-version-the-IDE-will-run-under).

**If you can try your IDEA to run on 1.8.0_51b it would be greatly appreciated.** On my Mac, 1.8.0_u51 runs fine, I can even include firebug-lite and get a semi-working debug panel. JDK 1.8.0_u60 crashes somewhere in the JavaFX libraries in native code so it core dumps and I can do nothing about debugging it. If your problem goes away when you run on the Oracle JDK 1.8.0_u51 then I can update the release notes and the code to fall back to the old preview if a version higher than that is used. At least until I can figure out why it is doing it.

**Comment by vsch** [comment]
@chunwei, @ghostsquad, All IntjelliJ products have their own way of configuring the boot JDK the products use. This may or may not actually get the product to boot with the JDK you expect.

Can you please verify the JDK that the product actually uses, not what it should be using based on your configuration? You can see the JDK in the about dialog.

I spent 3 hours trying to figure out why the plugin failed on PhpStorm 9.5 EAP only to notice that it ignored the configuration, which PhpStorm itself allowed me to select through its GUI. When I managed to get it to boot with 1.8.0_u51 the plugin worked.

Setting 'Use Old Preview' simply forces the code not to bother checking if JavaFX is available and always creates the old preview. Hence, no exceptions.

**Comment by chunwei** [comment]
@vsch on my win10 x64, webstorm 10.0.4, default bundle jre is 1.8.0.40 x86.
I have tried 1.8.0.60 amd64, and 1.8.0.51 amd64.   
All get the same error.  check screen captures below:
![2015-09-09](https://cloud.githubusercontent.com/assets/1955067/9751983/f0425836-56dd-11e5-945b-04bc97505828.png)
![2015-09-09 1](https://cloud.githubusercontent.com/assets/1955067/9751996/159a7d3e-56de-11e5-94e8-82ca64da7030.png)
![2015-09-09 2](https://cloud.githubusercontent.com/assets/1955067/9751997/15cf67e2-56de-11e5-99f5-b2e602c56903.png)

**Comment by vsch** [comment]
Thank you @chunwei. I will have to setup a Windows based IDEA CE environment and try to see what the issue is. I it something with either the way I initialize JavaFX but haven't figured it out yet. It is hard to mess up code that I took from Oracle examples, it is literally one line: `JFXPanel jfxPanel = new JFXPanel();. The error is coming from JavaFX code deep in the library there is a null pointer. I am perusing JavaFX in swing applications documentation now.

I will keep you posted on my progress.

**Comment by jdelic** [comment]
> If you can try your IDEA to run on 1.8.0_51b it would be greatly appreciated.

No problem. So here are my current results:

 Architecture  |          JDK version        | Freeze status | Run via                       |
:-------------:|----------------------------:|:-------------:|:------------------------------|
     x64       | JDK 1.8.0_51-b16 amd64      |  :red_circle: | idea64.exe                    |
     x86       | JRE 1.8.0_40-b26 (included) |  :red_circle: | idea.exe                      |
     x86       | JDK 1.8.0_60-b27            |  :red_circle: | IDEA_JDK environment variable |

As you can see I tested the current JDK version and the previous one which I had already installed. I verified the distinct JDK versions through IDEA's about dialog. I wasn't able to find any working configuration. Reproducing the bug is easy because *any edit* to a MarkDown document will immediately freeze the IDE (I think you have to insert at least one new-line into the document so that a WebView update is triggered).

It seems that the JavaFX WebView is *severely boinked*<sup>:tm:</sup>

**Comment by vsch** [comment]
New version released. Will run with JDK 1.6. May work with JDK 1.8 on Windows, various revisions. I think the issue for the latter was the url(data: ) embedded font in the CSS. Now moved to temporary file.

**Comment by jdelic** [comment]
I just tested version `1.1.1` of the plug-in with the JRE 1.8.0_40-b26 which comes with the IntelliJ IDEA installer and was unable to reproduce the exception (with "Use Old Preview" turned off in the settings, obviously).

**Comment by vsch** [comment]
New release available 1.1.2:
- Fix Exception reporting for ScratchFileService not found on older versions. If the check fails then no others will be performed until build of IDE changes.
- Fix WebView Preview Tab not working for 1.8u60 and up. In 1.8u60+ it no longer handles fonts with uri(data:) nor font/woff only .ttf.
- Fix allowing capital X to mark closed tasks. Fix #24
- Fix lack of ComboBox for Build 133.1711, now for earlier builds using List for theme selection instead of ComboBox.
- Change since-build to 133.1711, earliest tested version IDEA 13.0 CE, new Preview is not available but old preview works.

Please let me know if it solves your issue and/or close the issue which will be taken as issue solved.
